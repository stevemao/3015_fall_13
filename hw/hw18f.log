This is pdfTeXk, Version 3.141592-1.40.3 (Web2C 7.5.6) (format=pdflatex 2012.7.3)  26 NOV 2013 12:34
entering extended mode
 %&-line parsing enabled.
**hw18f
(./hw18f.tex
LaTeX2e <2005/12/01>
Babel <v3.8h> and hyphenation patterns for english, usenglishmax, dumylang, noh
yphenation, arabic, basque, bulgarian, coptic, welsh, czech, slovak, german, ng
erman, danish, esperanto, spanish, catalan, galician, estonian, farsi, finnish,
 french, greek, monogreek, ancientgreek, croatian, hungarian, interlingua, ibyc
us, indonesian, icelandic, italian, latin, mongolian, dutch, norsk, polish, por
tuguese, pinyin, romanian, russian, slovenian, uppersorbian, serbian, swedish, 
turkish, ukenglish, ukrainian, loaded.
(/usr/share/texmf/tex/latex/base/article.cls
Document Class: article 2005/09/16 v1.4f Standard LaTeX document class
(/usr/share/texmf/tex/latex/base/size11.clo
File: size11.clo 2005/09/16 v1.4f Standard LaTeX file (size option)
)
\c@part=\count79
\c@section=\count80
\c@subsection=\count81
\c@subsubsection=\count82
\c@paragraph=\count83
\c@subparagraph=\count84
\c@figure=\count85
\c@table=\count86
\abovecaptionskip=\skip41
\belowcaptionskip=\skip42
\bibindent=\dimen102
)
\@input{header.aux}
\openout2 = `header.aux'.

(./header.tex (/usr/share/texmf/tex/latex/amsfonts/amssymb.sty
Package: amssymb 2002/01/22 v2.2d

(/usr/share/texmf/tex/latex/amsfonts/amsfonts.sty
Package: amsfonts 2001/10/25 v2.2f
\@emptytoks=\toks14
\symAMSa=\mathgroup4
\symAMSb=\mathgroup5
LaTeX Font Info:    Overwriting math alphabet `\mathfrak' in version `bold'
(Font)                  U/euf/m/n --> U/euf/b/n on input line 132.
)) (./remark.sty)
(/home/faculty/jlc/papers/tex-stuff/bussproofs.sty
Proof Tree (bussproofs) style macros. Version 0.6c.
\theLevel=\count87
\myMaxLevel=\count88
\myBoxA=\box26
\myBoxB=\box27
\myBoxC=\box28
\myBoxD=\box29
\myBoxLL=\box30
\myBoxRL=\box31
\thisAboveSkip=\dimen103
\thisBelowSkip=\dimen104
\newScoreStart=\dimen105
\newScoreEnd=\dimen106
\newCenter=\dimen107
\displace=\dimen108
\leftLowerAmt=\dimen109
\rightLowerAmt=\dimen110
\scoreHeight=\dimen111
\scoreDepth=\dimen112
)
\c@theorem=\count89
\c@Rule=\count90
\c@conjecture=\count91
\c@corollary=\count92
\c@example=\count93
\c@fact=\count94
\c@lemma=\count95
\c@definition=\count96
\c@claim=\count97
\c@remark=\count98
\c@slogan=\count99
\c@axiom=\count100
\c@note=\count101
\c@problem=\count102
\c@exercise=\count103
) (./hw18f.aux)
\openout1 = `hw18f.aux'.

LaTeX Font Info:    Checking defaults for OML/cmm/m/it on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for T1/cmr/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for OT1/cmr/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for OMS/cmsy/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for OMX/cmex/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.
LaTeX Font Info:    Checking defaults for U/cmr/m/n on input line 5.
LaTeX Font Info:    ... okay on input line 5.

Underfull \hbox (badness 10000) in paragraph at lines 6--7

 []

LaTeX Font Info:    Try loading font information for U+msa on input line 31.
(/usr/share/texmf/tex/latex/amsfonts/umsa.fd
File: umsa.fd 2002/01/19 v2.2g AMS font definitions
)
LaTeX Font Info:    Try loading font information for U+msb on input line 31.

(/usr/share/texmf/tex/latex/amsfonts/umsb.fd
File: umsb.fd 2002/01/19 v2.2g AMS font definitions
)
Underfull \hbox (badness 10000) in paragraph at lines 49--51

 []


Underfull \hbox (badness 10000) in paragraph at lines 52--55

 []


Underfull \hbox (badness 10000) in paragraph at lines 56--59

 []

[1



{/var/lib/texmf/fonts/map/pdftex/updmap/pdftex.map}]
\myBox1=\box32
\myScoreStart1=\dimen113
\myCenter1=\dimen114
\myScoreEnd1=\dimen115
\myBox2=\box33
\myScoreStart2=\dimen116
\myCenter2=\dimen117
\myScoreEnd2=\dimen118

Overfull \hbox (30.21127pt too wide) in paragraph at lines 171--172
 [] 
 []

[2] [3]
Underfull \hbox (badness 10000) in paragraph at lines 266--271

 []


Underfull \hbox (badness 10000) in paragraph at lines 272--275

 []


Underfull \hbox (badness 10000) in paragraph at lines 276--279

 []


Underfull \hbox (badness 10000) in paragraph at lines 280--283

 []

[4] (./hw18f.aux) ) 
Here is how much of TeX's memory you used:
 924 strings out of 256216
 9321 string characters out of 1917073
 61290 words of memory out of 1500000
 4245 multiletter control sequences out of 10000+200000
 13569 words of font info for 54 fonts, out of 1200000 for 2000
 645 hyphenation exceptions out of 8191
 27i,10n,26p,238b,221s stack positions out of 5000i,500n,6000p,200000b,15000s
</usr/share/texmf/fonts/type1/bluesky/cm/cmbx10.pfb></usr/sh
are/texmf/fonts/type1/bluesky/cm/cmbx12.pfb></usr/share/texmf/fonts/type1/blues
ky/cm/cmmi10.pfb></usr/share/texmf/fonts/type1/bluesky/cm/cmmi8.pfb></usr/share
/texmf/fonts/type1/bluesky/cm/cmr10.pfb></usr/share/texmf/fonts/type1/bluesky/c
m/cmr7.pfb></usr/share/texmf/fonts/type1/bluesky/cm/cmr8.pfb></usr/share/texmf/
fonts/type1/bluesky/cm/cmsy10.pfb></usr/share/texmf/fonts/type1/bluesky/cm/cmsy
7.pfb></usr/share/texmf/fonts/type1/bluesky/cm/cmsy8.pfb></usr/share/texmf/font
s/type1/bluesky/cm/cmti10.pfb></usr/share/texmf/fonts/type1/bluesky/cm/cmtt10.p
fb>
Output written on hw18f.pdf (4 pages, 77414 bytes).
PDF statistics:
 63 PDF objects out of 1000 (max. 8388607)
 0 named destinations out of 1000 (max. 131072)
 1 words of extra memory for PDF output out of 10000 (max. 10000000)

