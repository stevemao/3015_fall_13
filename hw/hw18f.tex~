\documentclass[11pt]{article}

\include{header}

\begin{document}
\homework{18}{18 November}

\section{Type Inference}

Recall the type of terms.

\begin{program*}
\>  data Term = V String \\
\>            | Ap Term Term \\
\>            | Abs String Term \\
\end{program*}

\noindent{}The data-type {\tt{Type}} with products is:

\begin{program*}
\>   data Type = TyVar String | Arrow Type Type deriving Eq \\
\end{program*}

\subsection{Proof Rules}

Sequents in the system (which represent the state of a derivation) are of the form

Sequents in the system (which represent the state of a type derivation) are of
the form:

\[\Gamma,E \vdash M : T\]
In this structure, $\Gamma$ is a {\em{context}} representing a state of
knowledge about the types of some variables.  Contexts have the form:
 \[\Gamma = [x_1:\tau_1,\cdots{},x_k:\tau_k]\]
where the $x_i$'s are variables and $\tau_i$'s are types.

$E$ is a list of constraints between pairs
of types and in the rules is presented as follows:
\[E=\{\tau_{(1,1)} = \tau_{(1,2)}, \cdots , \tau_{(k,1)} = \tau_{(k,2)}\}\]
wher $\tau_{i,j}$'s are types.  

We write $\Gamma\backslash{}x$ to denote the list obtained from $\Gamma$ by
deleting all pairs whose first element is $x$.

As presented in the last homework, The proof rules for Wand's type
inference system are given as follows:


\AxiomRule{\Gamma,\{\alpha=\tau\}}{x:\tau}{Ax} {\hspace{.25in}} if $(x,\alpha)\in\Gamma$.
\vspace{.125in}\\

\SequentRule{\Gamma,E\cup{}\{\tau=\alpha\rightarrow\beta\}}{\lambda{}x.M : \tau}{[x:\alpha]{\tt{++}}(\Gamma\backslash{x}),\,E}{M:\beta}{Abs}
{\hspace{.25in}} where $\alpha$ and $\beta$ are fresh.
\vspace{.125in}\\

\SequentRuleTwo{\Gamma,E_1\cup{}E_2}{MN:\tau}{\Gamma,E_1}{M:\alpha\rightarrow\tau}{\Gamma,E_2}{N:\alpha}{App}
{\hspace{.25in}} where $\alpha$ is fresh.
\vspace{.125in}\\

A derivation in this system is a tree of instances of these rules where the
leaves of the tree are all instances of the (Ax) rule.  To construct a proof
that a closed term (no free variables) (say $M$) has a type, we postulate that
$M$ has some type (say $\alpha$) and proceed by recursion on the structure of
$M$ to show

\[ \exists{}E.[(Type,Type)]. {\rm{\ such \ that\ the \ sequent\ }}[],E\vdash M:\alpha {\rm{\  is \ derivable.}}\]

To find $E$, we use the proof rules above to try to construct a derivation
(leaving the $E$'s blank to start) and then propagate the constraints in the
$E$'s back down through the derivation tree from the leaves.

\begin{example}
Here is an example of a derivation that $\lambda{}x.x$ has a type by starting
with the sequent of the form  $[],\{??\}\vdash (\lambda{}x.x) :\tau$.  The term is an abstraction
so we apply the rule (Abs).

\begin{center}
\AxiomC{$[x:\alpha],E\vdash x : \beta$}
\LeftLabel{}\RightLabel{(Abs)}
\UnaryInfC{$[],\{\tau=\alpha \rightarrow \beta\}\cup{}E\vdash (\lambda{}x.x) :\tau$}
\DisplayProof
\end{center}

But if we fill in the set $E$ with the constraint $\tau=\alpha$, we have an
instance of the Axiom rule.


\begin{center}
\AxiomC{$E=\{\beta=\alpha\}$}
\LeftLabel{}\RightLabel{(Ax)}
\UnaryInfC{$[x:\alpha],E\vdash x : \beta$}
\LeftLabel{}\RightLabel{(Abs)}
\UnaryInfC{$[],\{\tau=\alpha \rightarrow \beta\}\cup{}E\vdash (\lambda{}x.x) :\tau$}
\DisplayProof
\end{center}

\noindent{}If we completely instantiate the sets $E$ we get the following complete derivation.

\begin{center}
\AxiomC{}
\LeftLabel{}\RightLabel{(Ax)}
\UnaryInfC{$[x:\alpha],\{\beta=\alpha\}\vdash x : \beta$}
\LeftLabel{}\RightLabel{(Abs)}
\UnaryInfC{$[],\{\tau=\alpha \rightarrow \beta,\beta=\alpha\}\vdash (\lambda{}x.x) :\tau$}
\DisplayProof
\end{center}

\end{example}


The fact that there is a derivation indicates that the term $(\lambda{}x.x)$
does have a type. We use the constraint set $E$ to actually determine the type of
$\lambda{}x.x$.  To do this, we unify the set $E$ and apply the resulting
substitution to the type $\tau$.  For this case, when we unify $E$ we get the
substitution $[\tau := \alpha\rightarrow\alpha, \beta:=\alpha]$.  Applying this
substitution to $\tau$ we determine that
$(\lambda{}x.x):\alpha\rightarrow\alpha$.



We can also do type derivations for terms containing free variables if we
assume those free variables do have types.

\begin{example}
Consider the term $y(\lambda{}x.x)$. This should have a type if
$y:(\alpha\rightarrow\alpha)\rightarrow\beta$.

We start by trying to show there is some $E$ such that there is a derivation of the sequent

\[[y:(\alpha\rightarrow\alpha)\rightarrow\beta],E\vdash y(\lambda{}x.x):\tau\]

\noindent{}Since the term is an application, we use the (Ap) rule.

\begin{center}
\AxiomC{$[y:(\alpha\rightarrow\alpha)\rightarrow\beta],E_1\vdash y :\alpha'\rightarrow\tau$}
\AxiomC{$[y:(\alpha\rightarrow\alpha)\rightarrow\beta],E_2\vdash (\lambda{}x.x):\alpha'$}
\LeftLabel{}\RightLabel{(Abs)}
\BinaryInfC{$[y:(\alpha\rightarrow\alpha)\rightarrow\beta],E_1\cup{}E_2\vdash y(\lambda{}x.x):\tau$}
\DisplayProof
\end{center}

\noindent{}The left branch is an instance of an axiom because there is an entry for the variable $y$ in the context.

\small{
\begin{center}
\AxiomC{$E_1 = \{\alpha'\rightarrow\tau=(\alpha\rightarrow\alpha)\rightarrow\beta\}$}
\LeftLabel{}\RightLabel{(Ax)}
\UnaryInfC{$[y:(\alpha\rightarrow\alpha)\rightarrow\beta],E_1\vdash y :\alpha'\rightarrow\tau$}
\AxiomC{$[y:(\alpha\rightarrow\alpha)\rightarrow\beta],E_2\vdash (\lambda{}x.x):\alpha'$}
\LeftLabel{}\RightLabel{(Abs)}
\BinaryInfC{$[y:(\alpha\rightarrow\alpha)\rightarrow\beta],E_1\cup{}E_2\vdash y(\lambda{}x.x):\tau$}
\DisplayProof
\end{center}
}

\noindent{}On the right branch we rebuild the proof given above.

\small{
\begin{center}
\AxiomC{$E_1 = \{\alpha'\rightarrow\tau=(\alpha\rightarrow\alpha)\rightarrow\beta\}$}
\LeftLabel{}\RightLabel{(Ax)}
\UnaryInfC{$[y:(\alpha\rightarrow\alpha)\rightarrow\beta],E_1\vdash y :\alpha'\rightarrow\tau$}
% \AxiomC{$[y:(\alpha\rightarrow\alpha)\rightarrow\beta],E_2\vdash (\lambda{}x.x):\alpha'$}
\AxiomC{$E_3=\{\beta'=\alpha''\}$}
\LeftLabel{}\RightLabel{(Ax)}
\UnaryInfC{$[x:\alpha'',y:(\alpha\rightarrow\alpha)\rightarrow\beta],E_3\vdash x : \beta'$}
\LeftLabel{}\RightLabel{(Abs)}
\UnaryInfC{$[y:(\alpha\rightarrow\alpha)\rightarrow\beta],E_2=(\{\alpha'=\alpha'' \rightarrow \beta'\}\cup{}E_3)\vdash (\lambda{}x.x) :\alpha'$}
\LeftLabel{}\RightLabel{(Abs)}
\BinaryInfC{$[y:(\alpha\rightarrow\alpha)\rightarrow\beta],E=(E_1\cup{}E_2)\vdash y(\lambda{}x.x):\tau$}
\DisplayProof
\end{center}
}

Putting together the constraints, we get the following set:

\[\begin{array}{lcl}
E & = & E_1 \cup E_2 \\
  & = & \{\alpha'\rightarrow\tau=(\alpha\rightarrow\alpha)\rightarrow\beta\} \cup (\{\alpha'=\alpha'' \rightarrow \beta'\}\cup{}E_3)\\
  & = & \{\alpha'\rightarrow\tau=(\alpha\rightarrow\alpha)\rightarrow\beta\} \cup (\{\alpha'=\alpha'' \rightarrow \beta'\}\cup{}\{\beta'=\alpha''\})\\
  & = & \{\alpha'\rightarrow\tau=(\alpha\rightarrow\alpha)\rightarrow\beta,\alpha'=\alpha'' \rightarrow \beta', \beta'=\alpha''\})\\
\end{array}\]

Unification of this results in the substitution:

\[s = [a' := (b' \rightarrow b'),t := b,a := b',a'' := b']\]

When $s$ is applied to $\tau$ we get the type $\beta$, as expected.

\end{example}



\subsection{Implementation}


In Haskell we encode contexts as list of type
{\tt{[(String,Type)]}}. Constraint sets are represented in the Haskell
implementaiton as a list of type {\tt{[(Type,Type)]}}. $M$ denotes a
lambda-term, and in Haskell is represented by elements of the data-type
{\tt{Term}}.  $T$ denotes a type and is represented in Haskell by elements of
the data-type {\tt{Type}}.


The implementation Here is the type of the {\tt{infer\_type}} function:
\begin{program*}
\> infer\_type :: [(String, Type)] \\ 
\>               -> Term \\ 
\>               -> Type \\ 
\>               -> [String] \\ 
\>               -> ([(Type, Type)], [String]) \\ 

\end{program*}
This function takes a context (denoted $\Gamma$ in the rules above and
represented by a list of {\tt{String}}, {\tt{Type}} pairs.), a term to infer
the type of, a type (denoted $\tau$ in the rules above and initially a type
variable not occurring anywhere in the context), and a string list containing
the names of all variables used so far.

\begin{program*}
\>  infer\_type context trm typ vars = \\ 
\>   case trm of \\ 
\>     (V x) ->  \\ 
\>        case (lookup x context) of  \\ 
\>          (Just a) -> ([(typ,a)],vars) \\ 
\>          Nothing ->  error ("infer\_type: " ++ x ++ " not in context!") \\ 
               
\>     (Ap m n) ->  \\ 
\>        let a = fresh "a" vars in \\ 
\>        let (e1,vars1) = infer\_type context m (Arrow (TyVar a) typ)(a:vars) in \\ 
\>        let (e2,vars2) = infer\_type context n (TyVar a) vars1 in \\ 
\>          (e1 ++ e2, vars2) \\ 

\>     (Abs x m) ->  \\ 
\>        let a = fresh "a" vars in \\ 
\>        let b = fresh "b" (a : vars) in \\ 
\>        let (e1,vars1) = infer\_type ((x,(TyVar a)):context) m (TyVar b) (a:b:vars) in \\ 
\>          ( [(typ, Arrow (TyVar a) (TyVar b))] ++ e1 , vars1) \\ 
\end{program*}

The case {\tt{V x}} implements the Axiom rule, the case labeled {\tt{(Ap m
n)}} implements the (Ap) rule and the case labeled {\tt{(Abs x m)}} implements
the (Abs) rule.


\subsection{Adding product types.}

To add product types  we extend the data-types {\tt{Type}} and {\tt{term}} as follows:

\begin{program*}
\>   data Type = TyVar String | Arrow Type Type | Prod Type Type \\
\> \\
\>  data Term = V String \\
\>            | Ap Term Term \\
\>            | Abs String Term \\
\>            | Spread Term (String,String) Term \\
\>            | Pair Term Term\\
\end{program*}

  Mathematically we write $M\times{}N$ for the Haskell term {\tt{Prod A B}} and
render the Haskell term {\tt{(Pair M N)}} as $\langle{}M,N,\rangle$ and we
write {\tt{(Spread M (x,y) N)}} as $\spread{M}{x,y}{N}$.
\ \\
Here are the additional proof rules:
\vspace{.125in}\\

\SequentRuleTwo{\Gamma,E_1\cup{}E_2\cup{}\{\tau=\alpha\times\beta\}}{\pair{M,N} : \tau}{\Gamma,E_1}{M:\alpha}{\Gamma,E_2}{N:\beta}{Pair}
{\hspace{.25in}} where $\alpha$ and $\beta$ are fresh.
\vspace{.125in}\\

\SequentRuleTwo{\Gamma,E_1\cup{}E_2}{\spread{M}{x,y}{N}:\tau}{\Gamma,E_1}{M:\alpha\times\beta}{\{x:\alpha,y:\beta\}\cup((\Gamma\backslash{}x)\backslash{}y),E_2}{N:\tau}{Spread}
{\hspace{.25in}}
\begin{tabular}{l}
 where $\alpha$ and \\
 $\beta$ are fresh.
\end{tabular}
\vspace{.125in}\\

\begin{exercise}
Using the base code provided (which includes unification for products) extend
the {\tt{infer\_type}} function to implement these additional type inference
rules.
\end{exercise}


\begin{exercise}
Design a number of test cases to show that your extension works.  You should at least include test for the following examples:
\[\begin{array}{ll}
\lambda{}f.\lambda{}x\lambda{}y.f\pair{x,y} & {\tt{(Abs \;\;''f'' (Abs \;\;''x'' (Abs \;\;''y'' (Ap (V \;\;''f'') (Pair (V \;\;''x'') (V \;\;''y''))))))}} \\
\lambda{}f.\lambda{}p.\spread{p}{x,y}{f\;\; x\;\; y} & {\tt{(Abs \;\;''f'' (Abs \;\;''p'' (Spread (V \;\;''p'') (''x'',''y'') (Ap (Ap (V \;\;''f'') (V \;\;''x''))  (V \;\;''y'')))))}} \\
\lambda{}f.\lambda{}p.\spread{p}{x,y}{f\;\; y\;\; x} & {\tt{(Abs \;\;''f'' (Abs \;\;''p'' (Spread (V \;\;''p'') (''x'',''y'') (Ap (Ap (V \;\;''f'') (V \;\;''y''))  (V \;\;''x'')))))}} \\
\lambda{}f.\lambda{}z\lambda{}w.f\pair{z,w}  & {\tt{(Abs \;\;''f'' (Abs \;\;''z'' (Abs \;\;''w'' (Ap  (V \;\;''f'') (Pair (V \;\;''z'') (V \;\;''w''))))))}} \\
\lambda{}p.\spread{p}{x,y}{x} & {\tt{(Abs \;\;''p'' (Spread (V \;\;''p'') (''x'',''y'') (V \;\;''x'')))}} \\
\lambda{}p.\spread{p}{x,y}{y} & {\tt{(Abs \;\;''p'' (Spread (V \;\;''p'') (''x'',''y'') (V \;\;''y'')))}} \\
\lambda{}p.\spread{p}{x,y}{\pair{y,x}} &{\tt{(Abs \;\;''p'' (Spread (V \;\;''p'') (''x'',''y'') (Pair (V \;\;''y'')(V \;\;''x''))))}}\\
\end{array}\]

\end{exercise}


\end{document}


% Local Variables:
% mode:latex
% comment-column:0
% comment-start: "\\> "
% comment-end: "\\\\ "
% compile-command: "pdflatex hw18f"
% fill-column:79
% End:




