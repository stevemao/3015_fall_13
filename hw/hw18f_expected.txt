
*TypeInf2> typeof (Abs "x" (Var "x"))
typeof (Abs "x" (Var "x"))
(a -> a)
*TypeInf2> typeof (Ap (Abs "x" (Var "x"))(Abs "x" (Var "x")))
typeof (Ap (Abs "x" (Var "x"))(Abs "x" (Var "x")))
(a -> a)
*TypeInf2> typeof (Abs "x" (Abs "y" (Ap (Var "y") (Var "x"))))
typeof (Abs "x" (Abs "y" (Ap (Var "y") (Var "x"))))
(a -> ((a -> b) -> b))
*TypeInf2> typeof (Abs "x" (Abs "y" (Abs "z" (Ap (Ap (Var "x") (Var "z"))(Ap (Var "y") (Var "z"))))))
typeof (Abs "x" (Abs "y" (Abs "z" (Ap (Ap (Var "x") (Var "z"))(Ap (Var "y") (Var "z"))))))
((a -> (b -> c)) -> ((a -> b) -> (a -> c)))
*TypeInf2> typeof (Fst (Pair (Abs "x" (Var "x")) (Abs "x" (Var "x"))))
typeof (Fst (Pair (Abs "x" (Var "x")) (Abs "x" (Var "x"))))
(a -> a)
*TypeInf2> typeof (Fst (Pair (Abs "x" (Abs "y" (Ap (Var "y") (Var "x")))) (Abs "x" (Var "x"))))
typeof (Fst (Pair (Abs "x" (Abs "y" (Ap (Var "y") (Var "x")))) (Abs "x" (Var "x"))))
(a -> ((a -> b) -> b))
*TypeInf2> typeof (Snd (Pair (Abs "x" (Abs "y" (Ap (Var "y") (Var "x")))) (Abs "x" (Var "x"))))
typeof (Snd (Pair (Abs "x" (Abs "y" (Ap (Var "y") (Var "x")))) (Abs "x" (Var "x"))))
(a -> a)
*TypeInf2> typeof (Snd (Pair (Abs "x" (Var "x")) (Abs "x" (Abs "y" (Ap (Var "y") (Var "x"))))))
typeof (Snd (Pair (Abs "x" (Var "x")) (Abs "x" (Abs "y" (Ap (Var "y") (Var "x"))))))
(a -> ((a -> b) -> b))
*TypeInf2> typeof (Abs "x" (Ap (Var "x") (Var "x")))
typeof (Abs "x" (Ap (Var "x") (Var "x")))
*** Exception: unify: Occurs check failure on a1.
*TypeInf2> 
